### PR TITLE
Use OS case sensitivity settings

### PR DESCRIPTION
**Pull request by yufeih**
This PR added `PathComparer`, `PathComparison` and `IsCaseSensitive` to `PathUtility`, and modifies existing comparison to use os path comparison.

Special files like `docfx.yml`, `TOC.md`, `index.md` are ~~always treated as case insensitive regardless of os sensitivity~~ **also treated using os path comparison**.

#2966 #2782

### CODE REVIEW

File: src/docfx/build/legacy/LegacyExtensions.cs [hunk 1]
```diff
@@ -27,12 +27,12 @@ public static string ToLegacyOutputPathRelativeToBaseSitePath(this Document doc,
         public static string ToLegacySiteUrlRelativeToBaseSitePath(this Document doc, Docset docset)
         {
             var legacyOutputFilePathRelativeToSiteBasePath = doc.SiteUrl;
-            if (legacyOutputFilePathRelativeToSiteBasePath.StartsWith($"/{docset.Config.SiteBasePath}", StringComparison.Ordinal))
+            if (legacyOutputFilePathRelativeToSiteBasePath.StartsWith($"/{docset.Config.SiteBasePath}", PathUtility.PathComparison))
             {
                 legacyOutputFilePathRelativeToSiteBasePath = Path.GetRelativePath(docset.Config.SiteBasePath, doc.SiteUrl.Substring(1));
             }
 
-            return PathUtility.NormalizeFile(doc.FilePath.EndsWith("index.md", StringComparison.OrdinalIgnoreCase) ? $"{legacyOutputFilePathRelativeToSiteBasePath}index" : legacyOutputFilePathRelativeToSiteBasePath);
```
Comment (herohua) [inline src/docfx/build/legacy/LegacyExtensions.cs]: >OrdinalIgnoreCase [](start = 96, length = 17)

this one should always ignore case?

Comment (yufeih) [reply src/docfx/build/legacy/LegacyExtensions.cs]: we need a strategy for handling special file names like _TOC_, _index_, _docfx_. This PR always uses case insensitive comparison for these special files, but we could use os specific comparison.

Comment (yufeih) [reply src/docfx/build/legacy/LegacyExtensions.cs]: I'm leaning towards using os specifc comparison for these special file names. Sometimes we need to probe files like TOC.md, index.md, this is problematic on linux if it is case insensitive.

Comment (herohua) [reply src/docfx/build/legacy/LegacyExtensions.cs]: makes sense. let's go with OS specific

---
In reply to: [200537310](https://github.com/dotnet/docfx/pull/3004#discussion_r200537310) [](ancestors = 200537310)

**Pull request merged**